### PR TITLE
feat(be,fe): recovery flow — prepare_delegation + delegation stamping + get_delegation (PR 7 of email-recovery stack)

### PR DESCRIPTION
**Pull request by sea-snake**
## Summary
Recovery flow on top of the two-phase DNSSEC architecture. Stacked on #3880 (setup flow).

- `email_recovery_prepare_delegation(dns_input, session_pk)` — anonymous. Same as setup-prepare plus a FE-generated session public key that the eventual delegation will be bound to.
- `email_recovery_get_delegation(nonce, session_key, expiration)` — query. After `RecoveryReady`, the FE fetches the `SignedDelegation`.
- Reverse `address → AnchorNumber` stable index (memory ID 24) for resolving the verified `From:` to an anchor at recovery time.
- `RecoveryReady` status variant carries `anchor_number` so the FE seeds its auth store directly.
- Recovery delegation principals recognised by `check_authorization` via a new `AuthorizationKey::EmailRecoveryAddress` variant.

## PR Stack
| # | PR | Description | Status |
|---|---|---|---|
| 0 | [#3836](https://github.com/dfinity/internet-identity/pull/3836) | Design doc | Open |
| 1 | [#3838](https://github.com/dfinity/internet-identity/pull/3838) | DNSSEC verifier scaffold | Open |
| 2 | [#3877](https://github.com/dfinity/internet-identity/pull/3877) | DKIM verifier (RFC 6376) | Open |
| 3 | [#3878](https://github.com/dfinity/internet-identity/pull/3878) | DMARC alignment (RFC 7489) | Open |
| 4 | [#3879](https://github.com/dfinity/internet-identity/pull/3879) | DoH fallback | Open |
| 5+6 | [#3880](https://github.com/dfinity/internet-identity/pull/3880) | Setup flow (storage + smtp_request) | Open |
| **7** | **this PR** | **Recovery flow (delegation)** | Open |
| 8 | [#3882](https://github.com/dfinity/internet-identity/pull/3882) | Frontend + feature flag | Open |
| 9 | [#3883](https://github.com/dfinity/internet-identity/pull/3883) | Deploy/upgrade scripts: dnssec_config + doh_config | Open |
| 10 | [#3884](https://github.com/dfinity/internet-identity/pull/3884) | Email-recovery UX overhaul | Open |

## Changes during review

- `RecoveryReady` gained an `anchor_number` field. The recovery flow already knows the anchor at smtp time (address → anchor reverse index) — surfacing it on the status payload lets the FE seed its auth store directly instead of running a recovery-phrase-keyed lookup against an email-recovery delegation. (See PR #3882's discussion.)
- `bind_credential_to_anchor` now refuses cross-anchor rebinds (`EmailRecoveryError::AddressAlreadyRegistered`); same-anchor rebinds remain idempotent.
- Recovery delegation principals are now recognised by `check_authorization`: `AuthorizationKey` gains an `EmailRecoveryAddress(String)` variant and the authz check derives the canister-sig principal from `H(salt || "email-recovery" || lowercase(address) || anchor)` for each bound credential. `activity_bookkeeping` updates `last_used` on the matched credential, and the daily/monthly stats counter gets an `email_recovery_counter`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by timothyaterton — APPROVED** [review]
✅ LGTM — ship it; two follow-up observations below.

Good, well-structured recovery flow. The anonymous `prepare_delegation` → SMTP → `submit_dkim_leaf` → `get_delegation` pipeline is clean, and the security-critical parts are solid: the recipient↔kind cross-check in `smtp.rs` prevents flow confusion, the seed derivation uses length-prefixed domain-separated components to avoid collisions, and the reverse-address index enforces one-anchor-per-address at the storage layer.

---

## Security

No blocking concerns.

**[FOLLOW-UP]** `calculate_email_recovery_seed` uses a single-byte length prefix (`blob.push(salt.len() as u8)`). The salt is 32 bytes today, so `as u8` is fine and matches the existing `calculate_delegation_seed` pattern for OpenID. But if this seed format is being designed fresh, a `u32` length prefix (like a proper TLV) would eliminate any future confusion if salt length ever changed. Very minor — noting for awareness, not action.

**[FOLLOW-UP]** `email_recovery_get_delegation` is a query, which is correct for latency but means it runs against a potentially stale certified state on replicas. This mirrors the existing `get_delegation` pattern for passkeys and `openid_get_delegation`, so it's consistent and not a regression. Just flagging that the same replica-trust assumptions from the design doc's §8.5 apply here.

## Scalability

No concerns.

The reverse-address index (`SHA-256(lowercase(address))` → `AnchorNumber`) is well-designed: fixed 32-byte keys, single stable-memory B-tree, O(1) lookup at recovery time. The `check_authorization` loop over `anchor.email_recovery` is bounded by the ≤1 credential constraint enforced at the API layer.

## Maintainability

**[FOLLOW-UP]** `encode_dns_name_lowercase` is duplicated between `prepare.rs` and `submit_leaf.rs` (identical implementations). Consider extracting it to a shared location in the `email_recovery` module or a DNS-utility module in a follow-up. Not blocking — the function is small and stable.

---

Clean integration of the recovery flow on top of the setup primitives. The `SnapshotKind` / `RecipientFlow` dispatch is readable, the `RecoveryOutcome` caching avoids re-derivation in `get_delegation`, and the archive operations (`AddEmailRecovery` / `RemoveEmailRecovery`) correctly avoid leaking the address. Ship it.

File: src/internet_identity/internet_identity.did [hunk 1]
```diff
@@ -805,6 +820,10 @@ type IdentityInfo = record {
     name : opt text;
     // The timestamp at which the anchor was created
     created_at : opt Timestamp;
+    // The email-recovery credential bound to this anchor, if any.
+    // Lets the FE render the recovery-email card's active vs.
+    // inactive state without a separate query.
+    email_recovery : opt EmailRecoveryCredential;
```
Comment (aterga) [inline src/internet_identity/internet_identity.did]: Let's already serve all the recovery emails. 

The FE should just pick the first one for now, since the UI only allows setting one recovery email.

```suggestion
    email_recovery : vec EmailRecoveryCredential;
```

**Review by aterga — APPROVED** [review]
**PR Review: #3881 (feat(be,fe): recovery flow — prepare_delegation + delegation stamping + get_delegation)**

**Overall assessment**: Solid, well-architected PR. No blockers or major security gaps. The implementation faithfully executes the design doc (#3836) for the recovery half of the email-recovery stack. It correctly separates setup vs. recovery flows, uses a nonce-based pending challenge map with strict TTL/eviction, enforces one-email-per-anchor via the new reverse index, derives a domain-separated recovery principal for authorization, and stamps delegations only after full DKIM+DMARC verification. The recipient/kind cross-check in `smtp_request` and the seed derivation + reverse-lookup path are security-critical and appear robust.

The two-phase DNSSEC path (skeleton in `prepare_delegation` → leaf in `submit_dkim_leaf`) is complex but justified: it keeps `smtp_request` payloads small while still allowing the canister to do the crypto verification itself. DoH fallback is cleanly gated by the allowlist. Generated FE bindings and integration tests look complete for this slice of the stack.

### No major security gaps
- **Authorization**: Recovery uses a derived canister-signature principal (`H(salt || "email-recovery" || lowercase(address) || anchor)`) checked in `authz_utils::check_authorization` (new `AuthorizationKey::EmailRecoveryAddress` variant). Cross-anchor rebinds are refused. Recipient/kind mismatch in `smtp_request` prevents flow confusion attacks.
- **Replay / nonce safety**: 64-bit entropy nonces with collision retry loop, 30 min TTL, bounded pending map (10k entries, oldest-first eviction). `recovery_seed_for_nonce` + cached `RecoveryOutcome` prevents re-derivation.
- **Credential binding**: Only after full verification; `bind_credential_to_anchor` + reverse index enforces 1:1 mapping.
- **Session key binding**: Length-capped in `prepare_delegation`; stored only in pending until stamped.
- **Open endpoints**: `prepare_delegation` and `smtp_request`/`smtp_request_validate` are intentionally unauthenticated, but gated by nonce existence + cross-checks. No credential leakage.
- **Query safety**: `get_delegation` and `status` are queries (as expected for II delegation pattern); signature verification happens client-side.

Everything aligns with II's existing delegation / canister-sig model.

### Blockers
None. This PR is merge-ready once the tiny follow-ups below are addressed (all easy) and the stack lands (especially #3882 FE + feature flag and #3883 deploy scripts).

### Easy-to-change improvements (actionable, low effort)
1. **Deduplicate `encode_dns_name_lowercase`** (and its decode counterpart).  
   It's in `email_recovery/dns.rs` and (per reviewer note) duplicated elsewhere in the DNSSEC/DKIM code. Extract to `crate::dnssec::util` (or a new `dns_wire` helper) and import. This was already flagged in review—fix before merge to avoid future drift.

2. **Clarify/fix seed prefix in `calculate_email_recovery_seed`** (in `smtp.rs`).  
   Reviewer already called out u32 vs u8 inconsistency. Make it a consistent single-byte domain separator (matching other II seeds) and update the comment + any tests. Prevents future collision surface.

3. **Query staleness consistency for `email_recovery_get_delegation`**.  
   Document (in `.did` + Rust docs) how the query behaves relative to other II queries (e.g. standard `get_delegation`). If it reads heap pending + sig store, confirm FE handles signature verification correctly (it should, given the `SignedDelegation` return type). Tiny doc + test tweak if needed.

4. **Cleanup dead-code / follow-up markers**.  
   `mod.rs` has `#[allow(dead_code)]` and notes about recovery bits still landing in a follow-up PR. Either remove the allow or finish the wiring before merge. Also add full Rustdoc to new public fns to match the Candid changes.

5. **Minor test polish**.  
   The new integration test (`email_recovery.rs`) and canister test helpers are excellent. Add one explicit test for the cross-anchor rebind refusal (already implemented in `bind_credential_to_anchor`) and one for DoH fallback path edge case. Zero runtime cost.

### Strategic recommendations
- **Rate limiting on recovery entry points**: `prepare_delegation` and `submit_dkim_leaf` are anonymous and could be spammed to bloat the pending map or trigger unnecessary DNSSEC work. II already has rate-limiting patterns elsewhere—consider applying a lightweight counter (or reuse existing) here. The 10k cap + eviction mitigates but doesn't eliminate the DoS vector.
- **Storage / upgrade hygiene**: New reverse index (memory ID 24), `EmailRecoveryCredential` on anchors, and `recovery_outcome` in pending are all reasonable. The deploy/upgrade scripts (#3883) must handle the new stable structures cleanly. Post-merge, monitor stable memory growth in production (email-recovery anchors will be larger).
- **Activity / stats symmetry**: `activity_bookkeeping` + new `email_recovery_counter` is a good start. Ensure all future anchor ops (remove, list, etc.) treat `EmailRecoveryAddress` exactly like `DeviceKey`/`OpenIdCredentialKey`.
- **UX / feature flag alignment**: This PR wires the backend flow; #3882 adds the FE piece + flag. Make sure the flag gates *all* new methods (including the anonymous ones) so we can roll out safely.
- **Long-term maintainability**: The two-phase DNSSEC design is clever but adds surface area. Future recovery methods (e.g. passkeys-as-recovery) should reuse the pending + nonce + status machinery rather than inventing new flows. The `smtp.rs` recipient dispatch + kind cross-check is a great pattern—keep it as the single source of truth for mailbox semantics.

**Verdict**: LGTM (after the 5 easy fixes above). This is a high-quality addition to II that meaningfully improves recovery UX without compromising the security model. The stack as a whole (design → DNSSEC scaffold → DKIM → this flow) is landing cleanly.

Thanks for the thorough work (and co-authors)! Happy to re-review once the nits are addressed or if you want a deeper dive on any specific file.

File: src/internet_identity/src/authz_utils.rs [hunk 2]
```diff
@@ -125,6 +125,24 @@ pub fn check_authorization(
...
+        let seed = crate::email_recovery::smtp::calculate_email_recovery_seed(
+            &credential.address,
+            anchor_number,
+        );
+        let public_key = crate::delegation::der_encode_canister_sig_key(seed.to_vec());
+        if caller == Principal::self_authenticating(public_key) {
+            return Ok((
+                anchor.clone(),
+                AuthorizationKey::EmailRecoveryAddress(credential.address.clone()),
+            ));
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/authz_utils.rs]: Fixed in e7414a932: guard the email-recovery branch on `storage.salt().is_some()` before calling `calculate_email_recovery_seed`. `stamp_recovery_delegation` already awaits `ensure_salt_set()` before issuing, so no recovery delegation can exist when the salt is unset — an unauthorized caller now falls through to a clean `AuthorizationError` instead of trapping.

File: src/internet_identity/internet_identity.did (hunk 1 above)
Comment (aterga) [reply src/internet_identity/internet_identity.did]: Fixed in e7414a932: widened `IdentityInfo.email_recovery` from `opt EmailRecoveryCredential` to `vec EmailRecoveryCredential`. Storage already used `Vec`; this just brings the candid surface in sync. Canister API still caps at one entry — FE picks the first to render the recovery-email card.

File: src/internet_identity/internet_identity.did [hunk 3]
```diff
@@ -507,6 +507,160 @@ type OpenIdPrepareDelegationResponse = record {
     anchor_number : UserNumber;
 };
 
+// Email-recovery types
+// ====================
+// See `docs/ongoing/email-recovery.md` for the full design. This PR
+// only ships the setup half (binding flow); the recovery half lands
+// next.
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/internet_identity.did:514]: Addressed in [#3896](https://github.com/dfinity/internet-identity/pull/3896) (commit [bfcfafd11](https://github.com/dfinity/internet-identity/pull/3896/commits/bfcfafd11)) — the "Email-recovery types" header comment now covers both halves, and the service-method section header was renamed from "setup half" to plain "Email-recovery protocol" with the two flows spelled out.

File: src/internet_identity/src/main.rs [hunk 4]
```diff
@@ -1398,6 +1408,226 @@ mod openid_api {
...
+/// - Authorization (`prepare_add` / `credential_remove` are
+///   authenticated; `status` is anonymous since the nonce is the
+///   only secret needed to poll).
+/// - Wall-clock injection (`now_secs` lifted out of
+///   `ic_cdk::api::time` so the inner functions stay testable).
+///
+/// The recovery half (`prepare_delegation` / `get_delegation` /
+/// `recover@id.ai` dispatch in `smtp_request`) is intentionally not
+/// part of this PR — it needs a stable reverse address index that
+/// is best landed on its own.
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/main.rs:1425]: Addressed in [#3896](https://github.com/dfinity/internet-identity/pull/3896) (commit [bfcfafd11](https://github.com/dfinity/internet-identity/pull/3896/commits/bfcfafd11)) — the `email_recovery_api` module doc now describes both flows (setup binding + recovery delegation) and the `register@id.ai` / `recover@id.ai` dispatch.
